sudachi/src/dic/lexicon_set.rs

```diff
@@ -62,42 +59,38 @@ impl<'a> LexiconSet<'a> {
     /// pos_offset: number of pos in the grammar
     pub fn append(
         &mut self,
-        lexicon: Lexicon<'a>,
+        mut lexicon: Lexicon<'a>,
         pos_offset: usize,
     ) -> Result<(), LexiconSetError> {
         if self.is_full() {
             return Err(LexiconSetError::TooManyDictionaries);
         }
-
+        lexicon.set_dic_id(self.lexicons.len() as u8);
         self.lexicons.push(lexicon);
         self.pos_offsets.push(pos_offset);
         Ok(())
     }
 
     /// Returns if dictionary capacity is full
     pub fn is_full(&self) -> bool {
-        self.lexicons.len() >= LexiconSet::MAX_DICTIONARIES
+        self.lexicons.len() >= MAX_DICTIONARIES
     }
 }
 
 impl LexiconSet<'_> {
-    /// Returns a list of word_id and length of words that matches given input
+    /// Returns iterator which yields all words in the dictionary, starting from the `offset` bytes
     ///
-    /// Searches user dictionary first and then system dictionary
-    pub fn lookup(&self, input: &[u8], offset: usize) -> SudachiResult<Vec<(u32, usize)>> {
-        let mut vs: Vec<(u32, usize)> = Vec::new();
-        for (did, user_lexicon) in self.lexicons.iter().enumerate().skip(1) {
-            vs.extend(
-                user_lexicon
-                    .lookup(input, offset)?
-                    .iter()
-                    .map(|(wid, l)| self.build_dictword_id(did, *wid).map(|dwid| (dwid, *l)))
-                    .collect::<Result<Vec<_>, _>>()?,
-            );
-        }
-        vs.extend(self.lexicons[0].lookup(input, offset)?);
-
-        Ok(vs)
+    /// Searches dictionaries in the reverse order: user dictionaries first and then system dictionary
```

Comment:
Does the order affect the analysis result?

Reply:
It should not affect the analysis in any order (it is weight-based), except if two nodes have exactly same weights. I am not sure if user dictionary order and their effect on the analysis is specified somewhere.